kimchi/src/circuits/expr.rs

```diff
@@ -475,9 +475,15 @@ impl<F: FftField> PolishToken<F> {
                 EndoCoefficient => stack.push(c.endo_coefficient),
                 Mds { row, col } => stack.push(c.mds[*row][*col]),
                 VanishesOnLast4Rows => stack.push(eval_vanishes_on_last_4_rows(d, pt)),
-                UnnormalizedLagrangeBasis(i) => stack.push(
-                    d.evaluate_vanishing_polynomial(pt) / (pt - d.group_gen.pow(&[*i as u64])),
-                ),
+                UnnormalizedLagrangeBasis(i) => {
+                    // Renormalize negative values to wrap around at domain size
+                    let i = if *i < 0 {
+                        ((*i as i64) + (d.size as i64)) as u64
```

Comment:
this is great!

Reply:
would still be nice to add tests related to this, off by ones scare me :)

Reply:
good idea